core/lib/contract_verifier/src/compilers/solc.rs

```diff
@@ -0,0 +1,168 @@
+use std::{collections::HashMap, path::PathBuf, process::Stdio};
+
+use anyhow::Context;
+use serde::{Deserialize, Serialize};
+use tokio::io::AsyncWriteExt;
+use zksync_queued_job_processor::async_trait;
+use zksync_types::contract_verification_api::{
+    CompilationArtifacts, SourceCodeData, VerificationIncomingRequest,
+};
+
+use super::{parse_standard_json_output, Source};
+use crate::{error::ContractVerifierError, resolver::Compiler};
+
+// Here and below, fields are public for testing purposes.
+#[derive(Debug)]
+pub(crate) struct SolcInput {
+    pub standard_json: StandardJson,
+    pub contract_name: String,
+    pub file_name: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct StandardJson {
+    pub language: String,
+    pub sources: HashMap<String, Source>,
+    settings: Settings,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Settings {
+    /// The output selection filters.
+    output_selection: Option<serde_json::Value>,
+    /// Other settings (only filled when parsing `StandardJson` input from the request).
+    #[serde(flatten)]
+    other: serde_json::Value,
+}
+
+#[derive(Debug)]
+pub(crate) struct Solc {
+    path: PathBuf,
+}
+
+impl Solc {
+    pub fn new(path: PathBuf) -> Self {
+        Self { path }
+    }
+
+    pub fn build_input(
+        req: VerificationIncomingRequest,
+    ) -> Result<SolcInput, ContractVerifierError> {
+        // Users may provide either just contract name or
+        // source file name and contract name joined with ":".
+        let (file_name, contract_name) =
+            if let Some((file_name, contract_name)) = req.contract_name.rsplit_once(':') {
+                (file_name.to_string(), contract_name.to_string())
+            } else {
+                (
+                    format!("{}.sol", req.contract_name),
+                    req.contract_name.clone(),
+                )
+            };
+        let default_output_selection = serde_json::json!({
+            "*": {
+                "*": [ "abi", "evm.bytecode", "evm.deployedBytecode" ],
+                 "": [ "abi", "evm.bytecode", "evm.deployedBytecode" ],
+            }
+        });
+
+        let standard_json = match req.source_code_data {
+            SourceCodeData::SolSingleFile(source_code) => {
+                let source = Source {
+                    content: source_code,
+                };
+                let sources = HashMap::from([(file_name.clone(), source)]);
+                let settings = Settings {
+                    output_selection: Some(default_output_selection),
+                    other: serde_json::json!({
+                        "optimizer": {
+                            "enabled": req.optimization_used,
+                        },
```

Comment:
AFAIU, contract verifiers like Etherscan commonly allow specifying the intended number of runs for the optimizer (200 by default), but this setting doesn't seem to be present in the verification DTOs rn. Again, I'd prefer to add its support (if necessary) separately since this PR is large as-is.

Reply:
It is certainly required, but indeed it makes sense to handle it separately.

Reply:
I'm not sure if we should keep `SolSingleFile` type of endpoint. Probably it's better to fully deprecate it, and let wrapping request for single-file happen on client side (hardhat plugin already doesn't use `SolSingleFile` for a long time, explorer uses it but it should be easy to switch, I'm not sure if it's used in foundry and etherscan-like API